chalk-integration/src/lowering.rs

```diff
@@ -96,110 +95,77 @@ enum ApplyTypeLookup {
-                return Err(RustIrError::IncorrectNumberOfTypeParameters {
-                    identifier: name.clone(),
-                    expected: k.binders.len(interner),
-                    actual: 0,
-                });
-            } else {
-                return Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
-                    name: chalk_ir::TypeName::Closure(*id),
-                    // See note in `program`. Unlike rustc, we store upvars separately.
-                    substitution: chalk_ir::Substitution::empty(interner),
-                })
-                .intern(interner)
-                .cast(interner));
+        match self.lookup_apply_type(name) {
+            Ok(ApplyTypeLookup::Param(p)) => {
+                let b = p.skip_kind();
+                match &p.kind {
+                    chalk_ir::VariableKind::Ty(_) => Ok(chalk_ir::TyData::BoundVar(*b)
+                        .intern(interner)
+                        .cast(interner)),
+                    chalk_ir::VariableKind::Lifetime => Ok(chalk_ir::LifetimeData::BoundVar(*b)
+                        .intern(interner)
+                        .cast(interner)),
+                    chalk_ir::VariableKind::Const(ty) => {
+                        Ok(b.to_const(interner, ty.clone()).cast(interner))
+                    }
+                }
             }
-        }
-
-        if let Some(id) = self.opaque_ty_ids.get(&name.str) {
-            return Ok(
-                chalk_ir::TyData::Alias(chalk_ir::AliasTy::Opaque(chalk_ir::OpaqueTy {
-                    opaque_ty_id: *id,
+            Ok(ApplyTypeLookup::Adt(id)) => apply(self.adt_kind(id), chalk_ir::TypeName::Adt(id)),
+            Ok(ApplyTypeLookup::FnDef(id)) => {
+                apply(self.fn_def_kind(id), chalk_ir::TypeName::FnDef(id))
+            }
+            Ok(ApplyTypeLookup::Closure(id)) => {
+                apply(self.closure_kind(id), chalk_ir::TypeName::Closure(id))
+            }
+            Ok(ApplyTypeLookup::Opaque(id)) => Ok(chalk_ir::TyData::Alias(
+                chalk_ir::AliasTy::Opaque(chalk_ir::OpaqueTy {
+                    opaque_ty_id: id,
                     substitution: chalk_ir::Substitution::empty(interner),
-                }))
-                .intern(interner)
-                .cast(interner),
-            );
-        }
-
-        if let Some(id) = self.foreign_ty_ids.get(&name.str) {
-            return Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
-                name: chalk_ir::TypeName::Foreign(*id),
-                substitution: chalk_ir::Substitution::empty(interner),
-            })
+                }),
+            )
             .intern(interner)
-            .cast(interner));
-        }
-
-        if let Some(_) = self.trait_ids.get(&name.str) {
-            return Err(RustIrError::NotStruct(name.clone()));
+            .cast(interner)),
+            Err(_) => {
+                if let Some(id) = self.foreign_ty_ids.get(&name.str) {
```

Comment:
So, how did this end up on this path (I would expect it to be returned in `lookup_apply_type`)? Is it just because these can't have generics?

Reply:
I guess so, since there's no `foreign_ty_kinds`. Possibly `lookup_apply_type` and `lookup_generic_arg` should just use their own lookups, but it's kind of awkward to have all these separate maps.